File: api/src/validators/states/tn.ts
// api/src/validators/states/tn.ts
//
// Tamil Nadu border-tax validator. Scope: SCRIPTED path only (web source,
// human handover) — TN is not in the fully-automated set, so run.ts rejects
// path=fullyAutomated for it.
//
// Ported knowledge (old repo tn.py / defaults / dates):
//   - Tax Mode: WEEKLY / MONTHLY / QUARTERLY — TN has NO DAYS option.
//   - taxUpto is ALWAYS portal-derived on TN (e.g. WEEKLY -> Tax From + 6
//     days), so any client value is stripped for every mode (empty string
//     tells the worker the portal owns the field). `duration` is therefore
//     ignored here.
//   - TN was in the old NO_SAME_DAY set; moot in practice since we never
//     compute taxUpto for TN, noted for the record.
//   - Defaults: KRISHNAGIRI entry. ⚠️ GUESS carried over from the old repo —
//     TN's district dropdown was never captured; KRISHNAGIRI (Hosur border,
//     the high-traffic Karnataka->TN cab entry) is a placeholder. A miss falls
//     back to the first district on the worker. Confirm on the live portal and
//     update this default. Checkposts don't track the district either, so
//     entryCheckpoint defaults to "" (worker first-option strategy).
//   - Permit ALL INDIA TOURIST PERMIT with CONTRACT CARRIAGE PERMIT fallback —
//     this drives the Permit Fee line on TN's tax table; change the default if
//     the fleet runs on contract-carriage permits. Service Type NOT
//     APPLICABLE.
//   - paymentMethod is NOT pinned (operator picks at the portal). Optional,
//     pass-through uppercase, default UPI.
//   - No Distance field on TN, so none is emitted.

import {
    type StateValidator,
    type ValidationResult,
    type FieldError,
    ok,
    fail,
    str,
    normalizeDate,
    normalizeTaxTime,
    isVehicleNumber,
    normalizeVehicleNumber,
} from "../types";

const VALID_TAX_MODES = new Set(["WEEKLY", "MONTHLY", "QUARTERLY"]);

const DEFAULTS = {
    mobileNumber: "",
    // ⚠️ Placeholder from the old repo (see header) — confirm on the live
    // portal.
    entryDistrict: "KRISHNAGIRI",
    // TN checkposts don't track the district — empty means "first option" on
    // the worker.
    entryCheckpoint: "",
    permitType: "ALL INDIA TOURIST PERMIT",
    permitTypeFallback: "CONTRACT CARRIAGE PERMIT",
    serviceType: "NOT APPLICABLE",
    paymentMethod: "UPI",
} as const;

export const tnValidator: StateValidator = {
    code: "TN",
    name: "Tamil Nadu",

    validate(raw: Record<string, unknown>): ValidationResult {
        const missing: string[] = [];
        const invalid: FieldError[] = [];

        const vehicleNumber = str(raw.vehicleNumber);
        const taxMode = str(raw.taxMode)?.toUpperCase();
        const taxFrom = normalizeDate(raw.taxFrom);
        const paymentMethod = str(raw.paymentMethod)?.toUpperCase();
        const requestId = str(raw.requestId);
        const driverId = str(raw.driverId);
        // Optional; the live Tax From is plain type="date" per the old repo,
        // but runtime detection decides — accept the field regardless.
        const taxTime = normalizeTaxTime(raw.taxTime);

        if (!vehicleNumber) missing.push("vehicleNumber");
        if (!taxMode) missing.push("taxMode");
        if (!requestId) missing.push("requestId");
        if (!driverId) missing.push("driverId");
        if (!taxFrom) {
            if (str(raw.taxFrom)) {
                invalid.push({
                    field: "taxFrom",
                    reason: "expected YYYY-MM-DD or dd/Mon/yyyy",
                });
            } else {
                missing.push("taxFrom");
            }
        }
        if (taxTime === undefined) {
            invalid.push({
                field: "taxTime",
                reason: 'expected 24h "HH:MM", e.g. "09:30" or "14:00"',
            });
        }

        if (vehicleNumber && !isVehicleNumber(vehicleNumber)) {
            invalid.push({
                field: "vehicleNumber",
                reason: "not a valid Indian registration number",
            });
        }
        if (taxMode && !VALID_TAX_MODES.has(taxMode)) {
            invalid.push({
                field: "taxMode",
                reason:
                    "Tamil Nadu supports WEEKLY, MONTHLY or QUARTERLY only " +
                    "(there is no DAYS option on the TN portal)",
            });
        }

        if (missing.length || invalid.length) {
            return fail("Tamil Nadu", missing, invalid);
        }

        return ok({
            state: "TAMIL NADU",
            stateCode: "TN",
            // Informational on the scripted path — the operator picks the real
            // method at the portal.
            paymentMethod: paymentMethod || DEFAULTS.paymentMethod,
            requestId: requestId!,
            driverId: driverId!,
            mobileNumber: str(raw.mobileNumber) ?? DEFAULTS.mobileNumber,
            vehicleNumber: normalizeVehicleNumber(vehicleNumber!),
            taxMode: taxMode!,
            taxFrom: taxFrom!,
            // ALWAYS portal-derived on TN: empty tells the worker to only read
            // it back.
            taxUpto: "",
            ...(taxTime ? { taxTime } : {}),
            entryDistrict: (
                str(raw.entryDistrict) ?? DEFAULTS.entryDistrict
            ).toUpperCase(),
            entryCheckpoint: (
                str(raw.entryCheckpoint) ?? DEFAULTS.entryCheckpoint
            ).toUpperCase(),
            permitType: str(raw.permitType) ?? DEFAULTS.permitType,
            permitTypeFallback:
                str(raw.permitTypeFallback) ?? DEFAULTS.permitTypeFallback,
            serviceType: str(raw.serviceType) ?? DEFAULTS.serviceType,
            // No distance field on TN.
        });
    },
};

File: worker/src/scripted/states/tn.py
# worker/src/scripted/states/tn.py
"""Tamil Nadu border-tax SCRIPTED runner (web source, human handover).

New to this repo — state knowledge ported from the old repo's
scripted/border_tax/tn.py + _handover_runner.

TN specifics (from the old runner):
  owner_info     Entry District default KRISHNAGIRI (validator).
                 ⚠️ GUESS carried over from the old repo — TN's district
                 dropdown was never captured; KRISHNAGIRI (Hosur border, the
                 high-traffic Karnataka->TN cab entry) is a placeholder. A
                 miss falls back to the first district in the list. Confirm
                 against the live portal and update the validator default.
                 Checkposts don't track the district (old "first_option"
                 strategy) — the shared helper's first-option fallback
                 carries it.
  vehicle_info   Vehicle Category set-if-empty (LPV is typically the sole
                 real option); Permit Type only if empty (default ALL INDIA
                 TOURIST PERMIT, fallback CONTRACT CARRIAGE PERMIT — this
                 drives the Permit Fee line on the tax table); Service Type
                 NOT APPLICABLE, set-if-empty with first-option fallback;
                 NO Distance field.
  tax_info       TN offers WEEKLY (2) / MONTHLY (4) / QUARTERLY (5) — there
                 is NO DAYS option, so Tax Upto is ALWAYS portal-derived
                 (fills_tax_upto is False for every valid TN mode; the
                 validator strips taxUpto). Visible-text match carries all
                 three (note TN MONTHLY=4, not the generic 3 — the value map
                 below is TN's own); a mode the RC doesn't get aborts with
                 abort_reason="tax_mode_not_offered_for_rc" + the offered
                 list. Tax From is plain type="date" (runtime-detected).
"""

from __future__ import annotations

from engine.pipeline import Phase
from lifecycle.status import Status

from ..handover import DEFAULT_NEGATIVE_PATTERNS, HandoverConfig, make_handover_phase
from ..wizard import (
    make_open_portal,
    make_tax_info,
    make_vehicle_info,
    owner_info,
    select_service,
)

open_portal = make_open_portal("TN", "tn")

vehicle_info = make_vehicle_info(
    set_category_if_empty=True,
    has_distance=False,
    service_set_if_empty=True,
    service_first_option_fallback=True,
    manual_has_category=True,
    manual_has_distance=False,
    manual_datetime_dates=False,
)

# TN's own option values (old tn.py): WEEKLY=2, MONTHLY=4, QUARTERLY=5.
_TN_TAX_MODE_VALUES = {"WEEKLY": "2", "MONTHLY": "4", "QUARTERLY": "5"}

tax_info = make_tax_info(
    state_code="TN",
    tax_mode_values=_TN_TAX_MODE_VALUES,
    list_offered_on_miss=True,
)

HANDOVER_CONFIG = HandoverConfig(
    state_name="Tamil Nadu",
    receipt_markers=[
        "government of tamil nadu",
        "checkpost tax e-receipt",
        "receipt no",
        "grand total",
    ],
    negative_patterns=list(DEFAULT_NEGATIVE_PATTERNS),
)

PHASES: list[Phase] = [
    Phase("open_portal", open_portal, enter_status=Status.AI_AGENT_STARTED, max_secs=150),
    Phase("select_service", select_service, max_secs=150),
    Phase("owner_info", owner_info, max_secs=300),
    Phase("vehicle_info", vehicle_info, max_secs=180),
    Phase("tax_info", tax_info, max_secs=180),
    make_handover_phase(HANDOVER_CONFIG),
]
